src/job_handler.rs

```diff
@@ -1,35 +1,12 @@
+use anyhow::Result;
+use async_trait::async_trait;
+
 use crate::job::Job;
 use crate::JobSuccessType;
-use crate::JobSuccessType::*;
-use crate::errors::{ErrorKind, Result};
 
 pub type JobHandlerResult = Result<JobSuccessType>;
 
-pub trait JobHandler: Send {
-    fn handle(&mut self, job: &Job) -> JobHandlerResult;
-    fn cloned(&mut self) -> Box<dyn JobHandler>;
-}
-
-impl<F> JobHandler for F
-    where F: FnMut(&Job) -> JobHandlerResult + Copy + Send + 'static
-{
-    fn handle(&mut self, job: &Job) -> JobHandlerResult {
-        self(job)
-    }
-    fn cloned(&mut self) -> Box<dyn JobHandler> {
-        Box::new(*self)
-    }
-}
-
-pub fn printer_handler(job: &Job) -> JobHandlerResult {
-    info!("handling {:?}", job);
-    Ok(Success)
-}
-
-pub fn error_handler(_: &Job) -> JobHandlerResult {
-    Err(ErrorKind::JobHandlerError(Box::new("a".parse::<i8>().unwrap_err())).into())
+#[async_trait]
+pub trait JobHandler: Send + Sync {
+    async fn perform(&self, job: &Job) -> JobHandlerResult;
```

Comment:
@msepga has a different approach [using a `BoxFuture`](https://github.com/pganalyze/sidekiq_server.rs/pull/8/files#diff-48d27698cc708c7efe770fd897e4885efaa0a15cae30d54936068603ba03bbd9R38). That fixed the lifetime issues I had as of the first commit on this branch, but when integrating it into our own code, the futures generated by other crates caused compile errors because they aren't `Sync`.

Reply:
In principle this should be equivalent to `Box<dyn Sync + ...>` given that `JobFuture` requires `Sync` here, though we can merge as-is regardless 👍

Reply:
I'm not sure what you mean by `JobFuture` -- there isn't a type like that around.

My understanding is that the `BoxFuture` version and my attempt in the first commit was running into issues because it required that the returned futures be `Sync`. This implementation only requires that the `JobHandler` is `Sync`.

Reply:
Typo, should be `JobHandler` in my comment 🙂 So, `BoxFuture` doesn't require the future to be `Sync`, the linked implementation only required the *handler* closure to be `Sync`:

```
With the definition of BoxFuture:

pub type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a, Global>>;

Looking at the type now:

Box<dyn Send + Sync + FnMut(Job) -> BoxFuture<'static, Result<JobSuccessType>>>
    | The handler itself must: |    | The BoxFuture must:                    |
    | - impl Send              |    |  - Be Send                             |
    | - impl Sync              |    |  - Be 'static                          |
    | - Take a `Job` argument  |    |  - Return Result<JobSuccessType>       |
    | - Return `BoxFuture<..>`----> |                                        |
    `--------------------------'    |  The BoxFuture *does not* have to be   |
                                    |  Sync                                  |
                                    |                                        |
                                    `----------------------------------------'
```

[Here's a playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4450576bbfefb76ab6afd00498e720c3) that illustrates how the handler can be sync without the returned future being sync. You'll notice that uncommenting line 25 will fail to compile, because the returned future doesn't implement `Sync`, even if the handler is required to do so.